### PR TITLE
Lazy IMBA Silencer

### PR DESCRIPTION
**Pull request by X-TheDark**
Dota 2 Silencer with proper interactions with IMBA abilities (Arcane Curse/Last Word);

Includes English skill descriptions...I might add the other languages, since it's all copy paste. But not sure if it's worth it, considering that I do want to IMBA'fy Silencer as soon as it's cleared to do so.

This was all done as part of a personal initiative to get Silencer into IMBA, as in my opinion, he will drastically alter the landscape of 10v10's (that and I like coding and felt like learning how this is done with Dota 2 Custom Games).

**Feel free to contact me on Discord or leave a message here if there are any issues/additions/changes that you wish me to resolve.**
**I have done full ability tests with only 3-4 heroes (WD, Enigma, Puck, Axe, Necro too, actually), but I can't think of any kind of errant ability that stands out as potentially broken due to this.**

## Changes to hero and skills
* Arcane Curse/Int Steal aura are unchanged in values

### Hero
* 3 intelligence gain per level
* 1.5 BAT

### Glaives of Wisdom intelligence to damage changes
* Vanilla : 35%/50%/65%/80%
* Lazy IMBA : 45%/70%/95%/120%

### Global Silence
* CD reduced from 130 to 120/110/100
* Duration increased from 4/5/6 to 5/7/9
* Cast point decreased from 0.3 to 0.2

**Differences from vanilla:**
* **Last Word** : applies it's damage **before** the cast is done (Pugna Ward Zap style) - this change was coordinated with Firetoad;
* **Arcane Curse**/**Last Word** : applies the duration extension and silence (respectively) for **toggleable spells when turned ON** - change coordinated with Firetoad; (Examples include, Necro's Death Pulse being Toggle ON);

**Bugs:**
* As seen with the Enigma fix, now that the unit can die before casting something, there is a possibility of lua errors;
* Witch Doctor's Voodoo Restoration receives damage after you toggle it off (due to it lingering and spending mana shortly after being turned off).

**Weird interactions:**
* Puck is still the ultimate enigma as his Phase Shift seems to be the only channeling spell in the game that DOESN'T trigger "OnAbilityEndChannel" event, unlike things like Black Hole/WD's ult.

### CODE REVIEW

File: game/dota_addons/dota_imba/scripts/npc/npc_heroes_custom.txt [hunk 1]
```diff
@@ -1611,26 +1613,6 @@
...
-		//"Ability1"					"imba_silencer_curse_of_the_silent"
-		//"Ability2"					"imba_silencer_glaives_of_wisdom"
-		//"Ability3"					"imba_silencer_last_word"
-		//"Ability4"					"imba_silencer_global_silence"
-		"Ability5"					"attribute_bonus_int"
-
-		"ProjectileModel"			"particles/units/heroes/hero_silencer/silencer_base_attack.vpcf"
-	}
-
-	//=================================================================================================================
```
Comment (X-TheDark) [inline game/dota_addons/dota_imba/scripts/npc/npc_heroes_custom.txt]: Repeat override in custom heroes, hence removed.

File: game/dota_addons/dota_imba/scripts/vscripts/modifier/modifier_imba_arcane_curse_debuff.lua [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	if IsServer() then
+		if ( not params.ability:IsItem() ) and ( params.unit == self.parent ) then
+			-- Only extend duration of Toggle abilities when they are turned on
+			-- OnAbilityExecuted is ran before the toggle completes, so 'true' = we are about to turn it off
+			if params.ability:IsToggle() and params.ability:GetToggleState() then
+				return
+			end
+			-- The main reason this is a LUA script...so I don't have to do black magic to extend every individual
+			-- curse modifier, in case we have multiple of them (since they stack)
+			self:SetDuration( self:GetRemainingTime() + self.penalty_duration, true )
```
Comment (X-TheDark) [inline game/dota_addons/dota_imba/scripts/vscripts/modifier/modifier_imba_arcane_curse_debuff.lua]: Stackable modifiers introduce alot of pain...and lua code so I don't have to flip sideways to track individual stacks.

File: game/dota_addons/dota_imba/scripts/vscripts/modifier/modifier_imba_silencer_int_steal.lua [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	return false
+end
+
+function modifier_imba_silencer_int_steal:RemoveOnDeath()
+	return false
+end
+
+function modifier_imba_silencer_int_steal:GetAttributes()
+	return MODIFIER_ATTRIBUTE_IGNORE_INVULNERABLE + MODIFIER_ATTRIBUTE_PERMANENT
+end
```
Comment (X-TheDark) [inline game/dota_addons/dota_imba/scripts/vscripts/modifier/modifier_imba_silencer_int_steal.lua]: MODIFIER_ATTRIBUTE_PERMANENT seems to do nothing for lua modifiers, as you still lose it when you die even with this defined.

**Comment by fcalife** [comment]
Again, some conflicts due to different versions used, so I'll be adding this manually. Thank you very much for your work! :)

**Pull request closed without merging**